### PR TITLE
Slot Ability Separation

### LINKED ISSUE #109: Define Ability Slot Connection Interface
**Description:**
With the new separation of slots and abilities, there needs to be a way to "glue" the two concepts together.

**Acceptance Criteria:**
* The interface should be intuitive and clean
* The interface should allow the user to attach properties that is only relevant to the specific slot-ability connection.
* The resulting data structure should be able to easily answer questions such as:
  * What slots are connected to a specific ability?
  * What abilities are connected to a specific slot?
  * What properties are on the specific connection?

### LINKED ISSUE #110: Implement Separation Logic in fillSlot Stage
**Description:**
Implement and architect the process that fills the relevant slot in the `fillSlot` stage. Given the incoming message data from the user, fill the best slot. We no longer have to decide which ability's slots to traverse and fill as all common slots will be shared.

_What this task is NOT:_
This task does not include utilizing/extracting the information that can be derived from this step onward. This is the responsibility of the `evaluate` stage to determine the best move forward.

**Acceptance Criteria:**
* The most relevant slot should be filled with the user information

### LINKED ISSUE #111: Re-implement the Evaluate stage relating to missing slot data
**Description:**
The `evaluate` stage will need to be changed to account for the new slot-to-ability interface. The implementation to find the next desired slot given a specific ability will need to be updated.

**Acceptance Criteria:**
* Maintain feature parity with the current v2.3 architecture
* No new evaluate algorithms should be introduced

### PR DESCRIPTION
**Pull request by howlowck**
Separated the concept of slot and abilities.  

This PR reaches feature parity from our previous versions, while giving us the possibility of reusing previous slot values.

closes #109, closes #110, closes #111

### CODE REVIEW

File: src/stages/evaluate.ts [hunk 1]
```diff
@@ -216,7 +219,11 @@ export default function evaluate<T, G>(
 /**
  * Find the next enabled and pending slot in the `focusedAbility` to be prompted
  */
-function findNextSlotToPrompt<T, G>(getState: () => WolfState, abilities: Ability<T, G>[]): SlotId | null {
+function findNextSlotToPrompt<T, G>(
+  getState: () => WolfState,
+  slots: Slot<G>[],
+  abilities: Ability<T, G>[]
```
Comment (kevinleung23) [inline src/stages/evaluate.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/evaluate.ts [hunk 2]
```diff
@@ -227,34 +234,31 @@ function findNextSlotToPrompt<T, G>(getState: () => WolfState, abilities: Abilit
...
 }
 
 /**
  * Check if there are any abilities with all enabled slots filled.
  */
-function getAbilitiesCompleted<T, G>(getState: () => WolfState, abilities: Ability<T, G>[]): string[] {
+function getAbilitiesCompleted<T, G>(
+  getState: () => WolfState,
+  slots: Slot<G>[],
+  abilities: Ability<T, G>[]
```
Comment (kevinleung23) [inline src/stages/evaluate.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/evaluate.ts [hunk 3]
```diff
@@ -273,6 +277,7 @@ function getAbilitiesCompleted<T, G>(getState: () => WolfState, abilities: Abili
  */
 function isAbilityCompletedByFilledSlotsOnCurrentTurn<T, G>(
   getState: () => WolfState,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/stages/evaluate.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/evaluate.ts [hunk 4]
```diff
@@ -305,6 +310,7 @@ function isAbilityCompletedByFilledSlotsOnCurrentTurn<T, G>(
 
 function getMissingSlotsOnSlotStatus<T, G>(
   getState: () => WolfState,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/stages/evaluate.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/evaluate.ts [hunk 5]
```diff
@@ -331,6 +337,7 @@ function getMissingSlotsOnSlotStatus<T, G>(
  */
 function getUnfilledSlots<T, G>(
   getState: () => WolfState,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/stages/evaluate.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/fillSlot.ts [hunk 6]
```diff
@@ -445,6 +483,7 @@ async function runRetryCheck<T, G>(
   dispatch: Dispatch,
   getState: () => WolfState,
   convoStorageLayer: G,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/stages/fillSlot.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/stages/fillSlot.ts [hunk 7]
```diff
@@ -513,13 +552,14 @@ async function runRetryCheck<T, G>(
 async function runRetry<T, G>(
   convoStorageLayer: G,
   getState: () => WolfState,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/stages/fillSlot.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

File: src/tests/helpers/index.ts [hunk 8]
```diff
@@ -8,6 +8,7 @@ export interface ConversationTurn<T> {
 
 export interface TestCase<T, G> {
   description: string,
+  slots: Slot<G>[],
   abilities: Ability<T, G>[],
```
Comment (kevinleung23) [inline src/tests/helpers/index.ts]: Parameter `slots` and `abilities` can be simplified to pass in `flow` which contains both. Simplifies parameter from 2 -> 1

**Review by kevinleung23 — CHANGES_REQUESTED** [review]
Can simplify parameters.

**Review by kevinleung23 — APPROVED** [review]

**Review by gcrev93 — APPROVED** [review]

**Review by dnastrain — APPROVED** [review]
:shipit:

**Comment by dnastrain** [comment]
http://flow.microsoft.com FTW :)

**Pull request merged**